Add complex dtype support to mul

tarun292 opened the pull request:
Summary: Adding support for complex dtypes in op_mul. In this diff i'm only adding support in the optimized path and not in the core portable kernels as that needs support in a bunch of the element-wise macro's that i'm not really familiar with. For now this unblocks our internal use cases.

Differential Revision: D73877440

manuelcandales commented on kernels/test/op_mul_test.cpp:
```diff
@@ -472,6 +514,19 @@ TEST_F(OpMulOutTest, BothScalarInputBroadcastTest) {
+    // Disabling this for portable mode for now because currently we only support
+    // complex dtypes in optimized mode. Portable mode needs support in elementwise_ops.h
+    // for complex dtypes, which we don't have yet.
+    //ET_FORALL_COMPLEXH_TYPES(TEST_ENTRY);
```
The fact that you are commenting out this line means none of the changes you added in this PR are being tested. I agree you shouldn't be blocked by adding support in the elementwise macros. You have two possible solutions:
a) Add your optimized-only tests under https://github.com/pytorch/executorch/tree/main/kernels/optimized/test
b) Enable complex in portable op_mul without messing around with the elementise macros, just add an if/else and add similar code to the one you added in the optimized op.
I think b) will be way easier

tarun292 replied:
@manuelcandales took approach b and updated the PR.

manuelcandales reviewed (changes requested):
Test your code changes either by doing a) or b)

tarun292 committed:
Add complex dtype support to mul (#10560)

Summary:

Adding support for complex dtypes in op_mul. In this diff i'm only adding support in the optimized path and not in the core portable kernels as that needs support in a bunch of the element-wise macro's that i'm not really familiar with. For now this unblocks our internal use cases.

Differential Revision: D73877440
